packages/python/m/github/actions/api.py

```diff
@@ -0,0 +1,145 @@
+import os
+import sys
+from inspect import signature
+from typing import Any, Callable, TypeVar
+
+from m.cli import run_main
+from m.cli.cli import default_issue_handler
+from m.core import Bad, Good, Issue, Res, issue, rw
+from m.log import logging_config
+from m.pydantic import KebabModel
+from pydantic import TypeAdapter
+from pydantic.fields import FieldInfo
+
+InputModel = TypeVar('InputModel', bound=KebabModel)
+OutputModel = TypeVar('OutputModel', bound=KebabModel)
+
+
+def load_step_inputs(model: type[InputModel]) -> Res[InputModel]:
+    """Load the GitHub Action inputs from the environment.
+
+    This function can be used provided that in the GitHub Action workflow we
+    set up:::
+
+        env:
+            INPUT_ARG_1: ${{ inputs.arg-1 }}
+
+    In this way the inputs will set in the environment which we can then read
+    in our main python function.
+
+    Args:
+        model: The class to create an instance of.
+
+    Returns:
+        A `OneOf` with the model or an issue.
+    """
+    model_data: dict[str, str] = {}
+    for key, env_val in os.environ.items():
+        if key.startswith('INPUT_'):
+            _, input_name = key.split('INPUT_')
+            model_data[input_name.lower()] = env_val
+    try:
+        return Good(TypeAdapter(model).validate_python(model_data))
+    except Exception as ex:
+        return issue('load_step_inputs_failure', cause=ex)
+
+
+def dump_step_outputs(outputs: OutputModel) -> Res[int]:
+    """Dump the outputs for a GitHub Action step.
+
+    Github does this if we write to the file where $GITHUB_OUTPUT is pointing.
+
+    Args:
+        outputs: The outputs to dump.
+
+    Returns:
+        A `OneOf` with 0 or an issue.
+    """
+    output_data = '\n'.join([
+        f'{key}={output_val}'
+        for key, output_val in outputs.model_dump(by_alias=True).items()
+    ])
+    output_file = os.environ.get('GITHUB_OUTPUT', 'NO_GITHUB_OUTPUT_FILE')
+    return rw.write_file(output_file, output_data, 'a')
+
+
+def _result_handler(outputs: OutputModel) -> None:
+    try:
+        res = dump_step_outputs(outputs)
+    except Exception as ex:
+        default_issue_handler(Issue('dump_step_outputs', cause=ex))
+        sys.exit(5)
```

Comment:
why not raise an exception instead?

Reply:
This is to interact with bash to get a different error code out of the program

```
$ m some_command
$ echo $?
5
```

In this way we may be able to see in Github action that number instead of just `1`. Need to make a note in the docs.